src/ddmd/dinterpret.d

```diff
@@ -2047,6 +2047,23 @@ public:
             assert(result.op == TOKstructliteral || result.op == TOKclassreference);
             return;
         }
+
+        // we probably have a nested delegate here.
+        // the thing to do is to iterate up the stack until we find
+        // a this that has a matching type.
+        // https://issues.dlang.org/showbug.cgi?id=16301
+        size_t stackOffset = ctfeStack.savedThis.dim;
+        while(stackOffset && (!result || !result.type.equals(e.type)))
```

Comment:
Isn't this essentially implementing a dynamic scoping approach instead of a lexical scoping one?

If ctfeStack is the *call stack* - this sounds incorrect in the general case.